pkg/metricsadapter/provider/resourcemetrics.go

```diff
@@ -0,0 +1,463 @@
+		}
+	}
+
+	// step 2. Query metrics from the filtered target clusters
+	var nodeMetrics []metrics.NodeMetrics
+	for _, clusterName := range targetClusters {
+		sci := r.informerManager.GetSingleClusterManager(clusterName)
+		if sci == nil {
+			klog.Errorf("Failed to get cluster(%s) manager", clusterName)
+			continue
+		}
+
+		metricsV1betaGVR := metricsv1beta1.SchemeGroupVersion.WithResource("nodes")
+		metricsList, err := sci.GetClient().Resource(metricsV1betaGVR).List(context.Background(), metav1.ListOptions{
+			LabelSelector: selector,
+		})
+		if err != nil {
+			klog.Errorf("Failed to list node metrics in cluster(%s): %v", clusterName, err)
+			continue
+		}
+		internalMetrics, err := metricsConvertV1beta1NodeToInternalNode(metricsList.Items...)
+		if err != nil {
+			return nil, err
+		}
+
+		nodeMetrics = append(nodeMetrics, internalMetrics...)
+	}
+
+	return nodeMetrics, nil
+}
+
+// GetPodMetrics queries metrics by the internal constructed pod
+func (r *ResourceMetricsProvider) GetPodMetrics(pods ...*metav1.PartialObjectMetadata) ([]metrics.PodMetrics, error) {
+	var podMetrics []metrics.PodMetrics
+
+	// In the previous step, we construct pods list with only one element.
+	if len(pods) != 1 {
+		return podMetrics, nil
+	}
+
+	// In the previous step, if query with label selector, the name will be set to empty
+	if pods[0].Name == "" {
+		namespace := pods[0].Labels[namespaceLabelInternal]
+		selectorStr := pods[0].Labels[labelSelectorLabelInternal]
+		return r.queryPodMetricsBySelector(selectorStr, namespace)
+	}
+
+	return r.queryPodMetricsByName(pods[0].Name, pods[0].Namespace)
+}
+
+// GetNodeMetrics queries metrics by the internal constructed node
+func (r *ResourceMetricsProvider) GetNodeMetrics(nodes ...*corev1.Node) ([]metrics.NodeMetrics, error) {
+	var nodeMetrics []metrics.NodeMetrics
+
+	// In the previous step, we construct node list with only one element.
+	if len(nodes) != 1 {
+		return nodeMetrics, nil
+	}
+
+	// In the previous step, if query with label selector, the name will be set to empty
+	if nodes[0].Name == "" {
+		selectorStr := nodes[0].Labels[labelSelectorLabelInternal]
+		return r.queryNodeMetricsBySelector(selectorStr)
+	}
+
+	return r.queryNodeMetricsByName(nodes[0].Name)
+}
+
+// PodLister is an internal lister for pods
+type PodLister struct {
+	namespaceSpecified string
+}
+
+// NewPodLister creates an internal new PodLister
+func NewPodLister() *PodLister {
+	return &PodLister{}
+}
+
+// List returns the internal constructed pod with label selector info
+func (p *PodLister) List(selector labels.Selector) (ret []runtime.Object, err error) {
+	klog.Infof("List query pods with selector: %v", selector.String())
+
+	podData := &v1.PartialObjectMetadata{
+		TypeMeta: v1.TypeMeta{},
+		ObjectMeta: v1.ObjectMeta{
+			Labels: map[string]string{
+				namespaceLabelInternal:     p.namespaceSpecified,
+				labelSelectorLabelInternal: selector.String(),
+			},
+		},
```

Comment:
If query metrics use FieldSelector, The function in file `vendor/sigs.k8s.io/metrics-server/pkg/api/pod.go:L105` will is not be `nil` and  the FieldSelector will not match ant pods.

Reply:
query metrics only support label selector, do not support FieldSelector.

Reply:
I tried, and the result will be empty if with FieldSelctor, and it is as expected.

Reply:
> I tried, and the result will be empty if with FieldSelctor, 

I don't think so. 

https://github.com/karmada-io/karmada/pull/3578#discussion_r1205095944

If you want to keep the node metrics ,  the result should't  be empty.

Reply:
Ok, I fixed, please check again, here is the test scripts
```sh
kubectl --context=karmada-apiserver get --raw /apis/metrics.k8s.io/v1beta1/nodes/member3-control-plane
kubectl --context=karmada-apiserver get --raw /apis/metrics.k8s.io/v1beta1/nodes/member3-control-plane-none
kubectl --context=karmada-apiserver top nodes
kubectl --context=karmada-apiserver get --raw /apis/metrics.k8s.io/v1beta1/nodes\?fieldSelector\=metadata\.name=member2-control-plane

kubectl --context=karmada-apiserver top pods -A
kubectl --context=karmada-apiserver top pod nginx-748c667d99-qt2z8
kubectl --context=karmada-apiserver top pod nginx-748c667d99-qt2z8 -nkarmada-system
kubectl --context=karmada-apiserver top pods -nlocal-path-storage
kubectl --context=karmada-apiserver get --raw /apis/metrics.k8s.io/v1beta1/namespaces/karmada-system/pods?fieldSelector\=metadata\.name=karmada-agent-54dcc8846f-2mhgl
```